src/LiveComponent/src/Attribute/AsLiveComponent.php

```diff
@@ -0,0 +1,123 @@
+
+/**
+ * @author Kevin Bond <kevinbond@gmail.com>
+ *
+ * @experimental
+ */
+#[\Attribute(\Attribute::TARGET_CLASS)]
+final class AsLiveComponent extends AsTwigComponent
+{
+    /**
+     * @internal
+     *
+     * @param string|object $classOrObject
+     *
+     * @return LivePropContext[]
+     */
+    public static function liveProps($classOrObject): \Traversable
+    {
+        foreach (self::propertiesFor($classOrObject) as $property) {
+            if ($attribute = $property->getAttributes(LiveProp::class)[0] ?? null) {
+                yield new LivePropContext($attribute->newInstance(), $property);
+            }
+        }
+    }
+
+    /**
+     * @internal
+     *
+     * @param string|object $classOrObject
+     */
+    public static function isActionAllowed($classOrObject, string $action): bool
+    {
+        foreach (self::attributeMethodsFor(LiveAction::class, $classOrObject) as $method) {
+            if ($action === $method->getName()) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    /**
+     * @internal
+     *
+     * @param string|object $classOrObject
+     *
+     * @return \ReflectionMethod[]
+     */
+    public static function beforeReRenderMethods($classOrObject): \Traversable
+    {
+        yield from self::attributeMethodsFor(BeforeReRender::class, $classOrObject);
+    }
+
+    /**
+     * @internal
+     *
+     * @param string|object $classOrObject
+     *
+     * @return \ReflectionMethod[]
+     */
+    public static function postHydrateMethods($classOrObject): \Traversable
+    {
+        yield from self::attributeMethodsFor(PostHydrate::class, $classOrObject);
+    }
+
+    /**
+     * @internal
+     *
+     * @param string|object $classOrObject
+     *
+     * @return \ReflectionMethod[]
+     */
+    public static function preDehydrateMethods($classOrObject): \Traversable
+    {
+        yield from self::attributeMethodsFor(PreDehydrate::class, $classOrObject);
+    }
+
+    /**
+     * @param string|object $classOrObject
+     *
+     * @return \ReflectionMethod[]
+     */
+    private static function attributeMethodsFor(string $attribute, $classOrObject): \Traversable
+    {
+        foreach ((new \ReflectionClass($classOrObject))->getMethods(\ReflectionMethod::IS_PUBLIC) as $method) {
+            if ($method->getAttributes($attribute)[0] ?? null) {
+                yield $method;
+            }
+        }
+    }
+
+    /**
+     * @param string|object $classOrObject
+     *
+     * @return \ReflectionProperty[]
+     */
+    private static function propertiesFor($classOrObject): \Traversable
+    {
+        $class = $classOrObject instanceof \ReflectionClass ? $classOrObject : new \ReflectionClass($classOrObject);
+
+        foreach ($class->getProperties() as $property) {
+            yield $property;
+        }
+
+        if ($parent = $class->getParentClass()) {
+            yield from self::propertiesFor($parent);
+        }
```

Comment:
To get all of the properties on a class, we need to manually do the recursion up the tree? Does that handle traits? What if a property is overridden in a child component? Is it reported twice?

Reply:
We had a report on Slack (with the existing code) that this algorithm (which I think you just copied from a different part of the code for this PR) double-counts LiveProps in a parent class. In the `live-demo`, they created a `ChildNotificationComponent` and made it extend `NotificationComponent` - https://github.com/weaverryan/live-demo/blob/main/src/Twig/Components/NotificationComponent.php and got:

> An exception has been thrown during the rendering of a template ("The field name "expanded" cannot be used by multiple LiveProp properties in a component. Currently, both "expanded" and "expanded" are trying to use it in "App\Twig\Components\ChildNotificationComponent". Try adding LiveProp(fieldName="somethingElse") for the "expanded" property to avoid this.").

So indeed, I think we may have some double counting :).

Reply:
Ah ok, so what do you think? First one wins?

Reply:
Hmm, probably if you literally have 2 properties that both have `LiveProp` it *should* have an error. But in the user's case, I think (though I have no tested) that they don't actually have duplicated properties - it's just that our system is double-counting the same ONE property (e.g. perhaps counting it once when it scans the actual class [even though it lives in the parent] and a 2nd time when it scans the parent).

But also: especially if we force an *object* to be passed in (not just a `$class`), is there really no built-in way to get all the properties of an object (including parent classes and traits)?

Reply:
Confirmed the issue and added a fix in 4d3c8e87cb1a1df72d78200d3a0ce7cb809e4c37. I took a _first one found_ wins approach. Let me know if you'd like me to add something more robust.

> But also: especially if we force an object to be passed in (not just a $class), is there really no built-in way to get all the properties of an object (including parent classes and traits)?

Not for private/protected properties I don't believe (we allow LiveProp's to be private/protected).